cmd/gateway-oss.go

```diff
@@ -0,0 +1,843 @@
+	if err != nil {
+		return lpi, ossToObjectError(traceError(err), bucket, object)
+	}
+
+	defer func() {
+		// always drain output (response body)
+		io.CopyN(ioutil.Discard, resp.Body, 512)
+		resp.Body.Close()
+	}()
+
+	var lupr oss.ListUploadedPartsResult
+	err = xml.NewDecoder(resp.Body).Decode(&lupr)
+	if err != nil {
+		return lpi, ossToObjectError(traceError(err), bucket, object)
+	}
+
+	return fromOSSClientListPartsInfo(lupr, partNumberMarker), nil
+}
+
+// AbortMultipartUpload aborts a ongoing multipart upload.
+func (l *ossObjects) AbortMultipartUpload(bucket, object, uploadID string) error {
+	bkt, err := l.Client.Bucket(bucket)
+	if err != nil {
+		return ossToObjectError(traceError(err), bucket, object)
+	}
+
+	err = bkt.AbortMultipartUpload(oss.InitiateMultipartUploadResult{
+		Bucket:   bucket,
+		Key:      object,
+		UploadID: uploadID,
+	})
+	if err != nil {
+		return ossToObjectError(traceError(err), bucket, object)
+	}
+	return nil
+}
+
+// CompleteMultipartUpload completes ongoing multipart upload and finalizes object.
+func (l *ossObjects) CompleteMultipartUpload(bucket, object, uploadID string, uploadedParts []completePart) (oi ObjectInfo, err error) {
+	bkt, err := l.Client.Bucket(bucket)
+	if err != nil {
+		return oi, ossToObjectError(traceError(err), bucket, object)
+	}
+
+	imur := oss.InitiateMultipartUploadResult{
+		Bucket:   bucket,
+		Key:      object,
+		UploadID: uploadID,
+	}
+	parts := make([]oss.UploadPart, len(uploadedParts))
+	for i, up := range uploadedParts {
+		parts[i] = oss.UploadPart{
+			PartNumber: up.PartNumber,
+			ETag:       up.ETag,
+		}
+	}
+	_, err = bkt.CompleteMultipartUpload(imur, parts)
+	if err != nil {
+		return oi, ossToObjectError(traceError(err), bucket, object)
+	}
+
+	return l.GetObjectInfo(bucket, object)
+}
+
+// SetBucketPolicies sets policy on bucket.
+// OSS supports three types of bucket policies:
+// oss.ACLPublicReadWrite: readwrite in minio terminology
+// oss.ACLPublicRead: readonly in minio terminology
+// oss.ACLPrivate: none in minio terminology
+func (l *ossObjects) SetBucketPolicies(bucket string, policyInfo policy.BucketAccessPolicy) error {
+	bucketPolicies := policy.GetPolicies(policyInfo.Statements, bucket)
+	if len(bucketPolicies) != 1 {
+		return traceError(NotImplemented{})
+	}
+
+	prefix := bucket + "/*" // For all objects inside the bucket.
+	for policyPrefix, bucketPolicy := range bucketPolicies {
+		if policyPrefix != prefix {
+			return traceError(NotImplemented{})
+		}
+
+		var acl oss.ACLType
+		switch bucketPolicy {
+		case policy.BucketPolicyNone:
+			acl = oss.ACLPrivate
+		case policy.BucketPolicyReadOnly:
+			acl = oss.ACLPublicRead
+		case policy.BucketPolicyReadWrite:
+			acl = oss.ACLPublicReadWrite
+		default:
+			return traceError(NotImplemented{})
+		}
+
+		err := l.Client.SetBucketACL(bucket, acl)
+		if err != nil {
+			return ossToObjectError(traceError(err), bucket)
+		}
+	}
+
+	return traceError(NotImplemented{})
```

Comment:
SetBucketPolicies will always return NotImplamented{}, even if it applied policy successfully.